File: raceEngine/tools/tracks.js
'use strict';

// Relecture des circuits dessines, hors service.
//
// Le moteur refuse de demarrer sur un dessin faux — c'est ce qu'on veut en
// production, mais pas au moment ou l'on dessine. Cet outil fait la meme lecture
// et dit la meme chose, sans rien lancer :
//
//   node tools/tracks.js            verifie tous les circuits et les resume
//   node tools/tracks.js --order    ce que ca donne sur un grand prix entier
//
// Il traduit surtout le dessin en chiffres : combien de pixels fait le tour, ou
// tombe la ligne, a quelle profondeur chaque boite se pose. C'est la seule facon
// de verifier qu'un trace fait bien ce qu'on croyait dessiner.

const path = require('path');
const fs = require('fs');

function loadShared(name) {
    const candidates = ['..', '../../frontEnd/static/js'];
    for (const dir of candidates) {
        const full = path.join(__dirname, dir, name);
        if (fs.existsSync(full)) return require(full);
    }
    throw new Error(`${name} introuvable (cherche dans : ${candidates.join(', ')})`);
}

const CFG = loadShared('physics-config.js');
const track = require('../track');

const args = process.argv.slice(2);
const SHOW_ORDER = args.includes('--order');

let dir;
let tracks;
try {
    dir = track.resolveTracksDir(path.join(__dirname, '..'));
    tracks = track.loadTracks(dir, CFG);
} catch (err) {
    console.error(`✗ ${err.message}`);
    process.exit(1);
}

console.log(`${tracks.length} circuit(s) dans ${dir}\n`);

let warned = 0;

for (const circuit of tracks) {
    let cfg;
    try {
        cfg = track.applyTrack(CFG, circuit);
    } catch (err) {
        console.error(`✗ ${err.message}`);
        process.exit(1);
    }

    const width = cfg.world.width;

    // Duree d'un tour a la vitesse de defilement : c'est la mesure qui parle,
    // bien plus qu'un nombre de pixels. Cinq tours font la course.
    const lapSeconds = width / cfg.speeds.roadPPS;

    console.log(`── ${circuit.name}  [${circuit.source}]`);
    console.log(`   tour     ${circuit.columns} colonnes = ${width} px`
        + `  (~${lapSeconds.toFixed(1)} s le tour, ~${(lapSeconds * cfg.race.laps).toFixed(0)} s la course)`);
    console.log(`   ligne    colonne ${circuit.finishColumn} = ${cfg.world.finishLineX} px`);
    console.log(`   piste    ${circuit.rows} rangees sur ${cfg.road.minY}..${cfg.road.maxY} de profondeur`);
    console.log(`   camera   approche a ${cfg.race.cameraApproachDistance} px de l'arrivee`);

    // Les boites sont regroupees par colonne : c'est ainsi qu'un pilote les
    // rencontre — un rideau a franchir, pas des boites eparpillees. Et leur
    // place dans le tour decide du rythme des objets, ce que le dessin ne dit
    // pas a l'oeil.
    const columns = new Map();
    for (const box of cfg.world.itemBoxes) {
        if (!columns.has(box.x)) columns.set(box.x, []);
        columns.get(box.x).push(box.y);
    }

    console.log(`   boites   ${cfg.world.itemBoxes.length} en ${columns.size} rideau(x)`);
    for (const [x, depths] of [...columns.entries()].sort((a, b) => a[0] - b[0])) {
        let after = x - cfg.world.finishLineX;
        if (after < 0) after += width;
        console.log(`            x=${x} (${Math.round((after / width) * 100)} % du tour apres la ligne)`
            + `  profondeurs ${depths.map(y => y.toFixed(1)).join(', ')}`);
    }

    // Les pipes, et surtout ce qu'ils laissent passer. Un trace se juge la :
    // un passage juste au-dessus du minimum se franchit, mais huit karts n'y
    // tiennent pas de front.
    if (cfg.world.pipes.length) {
        console.log(`   pipes    ${cfg.world.pipes.length} : `
            + cfg.world.pipes.map(p => `x=${p.x} y=${p.y.toFixed(1)}`).join('  |  '));

        const passage = track.narrowestPassage(cfg, cfg.world.pipes, width);
        console.log(`   passage  ${passage.free.toFixed(1)} de libre au plus etroit `
            + `(vers x=${Math.round(passage.x)}), minimum exige ${cfg.pipe.minPassageY}`);
    }

    for (const warning of circuit.warnings) {
        console.log(`   ⚠ ${warning}`);
        warned++;
    }
    console.log('');
}

if (SHOW_ORDER) {
    console.log(`Grand prix de ${CFG.grandPrix.races} manches :`);
    for (let round = 1; round <= CFG.grandPrix.races; round++) {
        console.log(`   manche ${round} — ${track.forRound(tracks, round).name}`);
    }
    console.log('');
}

console.log(warned
    ? `✓ ${tracks.length} circuit(s) lisible(s), ${warned} avertissement(s).`
    : `✓ ${tracks.length} circuit(s) lisible(s), rien a signaler.`);

File: raceEngineCpp/src/tools/tracks.cpp
#include "tools/tracks.hpp"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <map>
#include <string>
#include <vector>

#include "track.hpp"

namespace tools {

namespace {

// Les nombres du JS s'ecrivent sans decimale inutile (`7680`, pas `7680.0`) :
// la sortie doit se comparer telle quelle a celle de l'outil node.
std::string num(double v) {
    if (v == std::floor(v) && std::abs(v) < 1e15) {
        return std::to_string(static_cast<long long>(v));
    }
    char buf[32];
    std::snprintf(buf, sizeof(buf), "%g", v);
    return buf;
}

std::string fixed(double v, int decimals) {
    char buf[64];
    std::snprintf(buf, sizeof(buf), "%.*f", decimals, v);
    return buf;
}

} // namespace

int run_tracks(const config::Config& cfg, bool showOrder) {
    std::string dir;
    std::vector<track::Track> tracks;

    try {
        dir = track::resolve_tracks_dir(".");
        tracks = track::load_tracks(dir, cfg);
    } catch (const std::exception& err) {
        std::printf("✗ %s\n", err.what());
        return 1;
    }

    std::printf("%zu circuit(s) dans %s\n\n", tracks.size(), dir.c_str());

    int warned = 0;

    for (track::Track& circuit : tracks) {
        config::Config applied;
        try {
            applied = track::apply_track(cfg, circuit);
        } catch (const std::exception& err) {
            std::printf("✗ %s\n", err.what());
            return 1;
        }

        const double width = applied.world.width;

        // Duree d'un tour a la vitesse de defilement : c'est la mesure qui
        // parle, bien plus qu'un nombre de pixels.
        const double lapSeconds = width / applied.speeds.roadPPS;

        std::printf("── %s  [%s]\n", circuit.name.c_str(), circuit.source.c_str());
        std::printf("   tour     %d colonnes = %s px  (~%s s le tour, ~%s s la course)\n",
                    circuit.columns, num(width).c_str(),
                    fixed(lapSeconds, 1).c_str(),
                    fixed(lapSeconds * applied.race.laps, 0).c_str());
        std::printf("   ligne    colonne %d = %s px\n",
                    circuit.finishColumn, num(applied.world.finishLineX).c_str());
        std::printf("   piste    %d rangees sur %s..%s de profondeur\n",
                    circuit.rows, num(applied.road.minY).c_str(), num(applied.road.maxY).c_str());
        std::printf("   camera   approche a %s px de l'arrivee\n",
                    num(applied.race.cameraApproachDistance).c_str());

        // Les boites sont regroupees par colonne : c'est ainsi qu'un pilote les
        // rencontre — un rideau a franchir, pas des boites eparpillees.
        std::map<double, std::vector<double>> columns;
        for (const config::Placed& box : applied.world.itemBoxes) {
            columns[box.x].push_back(box.y);
        }

        std::printf("   boites   %zu en %zu rideau(x)\n",
                    applied.world.itemBoxes.size(), columns.size());
        for (const auto& entry : columns) {
            double after = entry.first - applied.world.finishLineX;
            if (after < 0) after += width;

            std::string depths;
            for (size_t i = 0; i < entry.second.size(); i++) {
                if (i) depths += ", ";
                depths += fixed(entry.second[i], 1);
            }
            std::printf("            x=%s (%lld %% du tour apres la ligne)  profondeurs %s\n",
                        num(entry.first).c_str(),
                        static_cast<long long>(std::llround((after / width) * 100)),
                        depths.c_str());
        }

        // Les tuyaux, et surtout ce qu'ils laissent passer. Un trace se juge
        // la : un passage juste au-dessus du minimum se franchit, mais tout un
        // peloton n'y tient pas de front.
        if (!applied.world.pipes.empty()) {
            std::string list;
            std::vector<std::pair<double, double>> flat;
            for (size_t i = 0; i < applied.world.pipes.size(); i++) {
                const config::Placed& p = applied.world.pipes[i];
                if (i) list += "  |  ";
                list += "x=" + num(p.x) + " y=" + fixed(p.y, 1) + " "
                     + (p.red ? "red" : "green");
                flat.emplace_back(p.x, p.y);
            }
            std::printf("   pipes    %zu : %s\n", applied.world.pipes.size(), list.c_str());

            const track::Passage passage = track::narrowest_passage(applied, flat, width);
            std::printf("   passage  %s de libre au plus etroit (vers x=%lld), "
                        "minimum exige %s\n",
                        fixed(passage.free, 1).c_str(),
                        static_cast<long long>(std::llround(passage.x)),
                        num(applied.pipe.minPassageY).c_str());
        }

        for (const std::string& warning : circuit.warnings) {
            std::printf("   ⚠ %s\n", warning.c_str());
            warned++;
        }
        std::printf("\n");
    }

    if (showOrder) {
        std::printf("Grand prix de %d manches :\n", cfg.grandPrix.races);
        for (int round = 1; round <= cfg.grandPrix.races; round++) {
            std::printf("   manche %d — %s\n", round,
                        track::for_round(tracks, round).name.c_str());
        }
        std::printf("\n");
    }

    if (warned) {
        std::printf("✓ %zu circuit(s) lisible(s), %d avertissement(s).\n",
                    tracks.size(), warned);
    } else {
        std::printf("✓ %zu circuit(s) lisible(s), rien a signaler.\n", tracks.size());
    }
    return 0;
}

} // namespace tools
